### ISSUE TITLE
[native-image] Agent fails to detect reflective instantiation of arrays

### ISSUE THREAD

**Issue by lbialy** [oracle/graal#1794]
Hello,

During work on getting HikariCP working with native-image I have found out a problem with reflection configuration generated by native-image-agent. It seems that usage of 
`java.lang.reflect.Array.newInstance` 
method is not noticed by it and that it results with native-image successfully building binary that throws exception in runtime about unregistered array of concrete type.

Here's a piece of code allowing reproduction:
```java
import java.lang.reflect.Array;

class Bag<T> {
    private T[] elems;
    Bag(Class<?> clazz) {
        elems = (T[]) Array.newInstance(clazz, 3);
    }
}

class Elem {}

public class Main {
    public static void main(String[] args) {
        Bag<Elem> bag = new Bag<Elem>(Elem.class);
        System.out.println("Yay " + bag);
    }
}
```
I used maven to build this and package it into a jar. I then executed following command:
`java -agentlib:native-image-agent=config-output-dir=src/main/resources/META-INF/native-image/ -cp target/agent-misses-array-newinstance-1.0-SNAPSHOT.jar Main`, repackaged the jar, checked it for existence of `META-INF/native-image` path (it was there) and then built native-image of this jar successfully. Executing resulting binary yields:

```
Exception in thread "main" java.lang.IllegalArgumentException: Class Elem[] is instantiated reflectively but was never registered. Register the class by using org.graalvm.nativeimage.hosted.RuntimeReflection
        at com.oracle.svm.core.genscavenge.graal.AllocationSnippets.checkArrayHub(AllocationSnippets.java:182)
        at Main.main(Main.java:14)
```

Contents of `reflect-config.json`:
```json
[
]
```

Manually adding this information to `reflect-config.json` resolves the issue:
```json
  {
    "name": "Elem[]",
    "allDeclaredConstructors" : true,
    "allPublicConstructors" : true,
    "allDeclaredMethods" : true,
    "allPublicMethods" : true,
    "allDeclaredClasses" : true,
    "allPublicClasses" : true
  }
```

**Comment by peter-hofer** [comment]
Thank you @lbialy , I was able to reproduce the bug using your code snippet and will be looking into a fix.

**Comment by peter-hofer** [comment]
Fixed in b82943ffac1600a0488a2153f53c9ba31bb5014a. Thanks!
